\chapter{Using SPARQL with other programming languages}
\label{chap:programming}

\section{Using SPARQL with R}
\label{sec:sparql-with-r}

  Before we can start, we need to install the \texttt{SPARQL} package from
  \href{https://cran.r-project.org/web/packages/SPARQL/index.html}{CRAN}.

\begin{siderules}
\begin{verbatim}
install.packages("SPARQL")
\end{verbatim}
\end{siderules}

  Once the package is installed, we can load it:

\begin{siderules}
\begin{verbatim}
library("SPARQL")
\end{verbatim}
\end{siderules}

  Let's define where to send the query to:
\begin{siderules}
\begin{verbatim}
endpoint <- "http://localhost:8890/sparql"
\end{verbatim}
\end{siderules}

  $\ldots{}$ and the query itself:
\begin{siderules}
\begin{verbatim}
query <- "PREFIX vcf2rdf: <http://sparqling-genomics/vcf2rdf/>
PREFIX vc:    <http://sparqling-genomics/vcf2rdf/VariantCall/>
PREFIX rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX faldo: <http://biohackathon.org/resource/faldo#>

SELECT DISTINCT ?variant ?sample ?chromosome ?position ?filter
FROM <graph-name>
WHERE
{
  ?variant  rdf:type                vcf2rdf:VariantCall ;
            vcf2rdf:sample          ?sample ;
            faldo:reference         ?chromosome ;
            faldo:position          ?position ;
            vc:FILTER               ?filter .
}
LIMIT 10";
\end{verbatim}
\end{siderules}

  To actually execute the query, we can use the \texttt{SPARQL} function:
\begin{siderules}
\begin{verbatim}
query_data <- SPARQL (endpoint, query)
\end{verbatim}
\end{siderules}

  If the query execution went fine, we can gather the resulting
  \texttt{dataframe} from the \texttt{results} index.

\begin{siderules}
\begin{verbatim}
query_results <- query_data$results
\end{verbatim}
\end{siderules}

\subsection{Querying with authentication}

  When the SPARQL endpoint we try to reach requires authentication before
  it accepts a query, we can use the \texttt{curl\_args} parameter of the
  \texttt{SPARQL} function.

  In the following example, we use \texttt{dba} as username, and
  \texttt{secret-password} as password.

\begin{siderules}
\begin{verbatim}
endpoint     <- "http://localhost:8890/sparql-auth"
auth_options <- curlOptions(userpwd="dba:secret-password")
query        <- "SELECT DISTINCT ?p WHERE { ?s ?p ?o }"
query_data   <- SPARQL (endpoint, query, curl_args=auth_options)
results      <- query_data$results
\end{verbatim}
\end{siderules}

\section{Using SPARQL with GNU Guile}
\label{sec:sparql-with-guile}

  For Schemers using GNU Guile, the \href{https://github.com/roelj/guile-sparql}%
  {guile-sparql}\footnote{\url{https://github.com/roelj/guile-sparql}} package
  provides a SPARQL interface.

  The package provides a \texttt{driver} module that communicates with the
  SPARQL endpoint, a \texttt{lang} module to construct SPARQL queries using
  S-expressions, and a \texttt{util} module containing convenience functions.

  After installation, the modules can be loaded using:

\begin{siderules}
\begin{verbatim}
(use-modules (sparql driver)
             (sparql lang)
             (sparql util))
\end{verbatim}
\end{siderules}

  Using the \texttt{sparql-query} function, we can execute a query:

\begin{siderules}
\begin{verbatim}
(let ((endpoint       "http://localhost:8890/sparql-auth")
      (authentication "dba:secret-password")
      (query          "SELECT DISTINCT ?p WHERE { ?s ?p ?o }"))
  (display-query-results-of
    (sparql-query query
                  #:uri    endpoint
                  #:digest authentication)))
\end{verbatim}
\end{siderules}
